// @project-ajax/sdk/dist/capabilities/tool.js
import { Ajv } from "ajv";
import { createCapabilityContext } from "./context.js";
class InvalidToolInputError extends Error {
  constructor(message) {
    super(message);
    this.name = "InvalidToolInputError";
  }
  toJSON() {
    return {
      name: this.name,
      message: this.message
    };
  }
}
class InvalidToolOutputError extends Error {
  constructor(message) {
    super(message);
    this.name = "InvalidToolOutputError";
  }
  toJSON() {
    return {
      name: this.name,
      message: this.message
    };
  }
}
class ToolExecutionError extends Error {
  constructor(message) {
    super(message);
    this.name = "ToolExecutionError";
  }
  toJSON() {
    return {
      name: this.name,
      message: this.message
    };
  }
}
function createToolCapability(key, config) {
  const ajv = new Ajv();
  const validateInput = ajv.compile(config.schema);
  const validateOutput = config.outputSchema ? ajv.compile(config.outputSchema) : null;
  return {
    _tag: "tool",
    key,
    config: {
      title: config.title,
      description: config.description,
      schema: config.schema,
      outputSchema: config.outputSchema
    },
    async handler(input) {
      if (!validateInput(input)) {
        if (validateInput.errors == null) {
          throw new Error(
            "Unexpected: No validation errors after failed validation"
          );
        }
        const result = {
          _tag: "error",
          error: new InvalidToolInputError(
            JSON.stringify(validateInput.errors, null, 2)
          )
        };
        process.stdout.write(`
<output>${JSON.stringify(result)}</output>
`);
        return result;
      }
      try {
        const capabilityContext = createCapabilityContext();
        const result = await config.execute(input, capabilityContext);
        if (validateOutput && !validateOutput(result)) {
          const result2 = {
            _tag: "error",
            error: new InvalidToolOutputError(
              JSON.stringify(validateOutput.errors, null, 2)
            )
          };
          process.stdout.write(
            `
<output>${JSON.stringify(result2)}</output>
`
          );
          return result2;
        }
        process.stdout.write(`
<output>${JSON.stringify(result)}</output>
`);
        return {
          _tag: "success",
          value: result
        };
      } catch (err) {
        const result = {
          _tag: "error",
          error: new ToolExecutionError(
            err instanceof Error ? err.message : String(err)
          )
        };
        process.stdout.write(`
<output>${JSON.stringify(result)}</output>
`);
        return result;
      }
    }
  };
}
export {
  InvalidToolInputError,
  InvalidToolOutputError,
  ToolExecutionError,
  createToolCapability
};
