// @project-ajax/sdk/dist/builder.js
function richText(content) {
  return [[content]];
}
function url(url2) {
  return [[url2]];
}
function title(content) {
  return [[content]];
}
function text(content) {
  return [[content]];
}
function email(email2) {
  return [[email2]];
}
function phoneNumber(phone) {
  return [[phone]];
}
function checkbox(checked) {
  return checked ? [["Yes"]] : [["No"]];
}
function file(fileUrl, fileName) {
  return [[fileName ?? fileUrl, [["a", fileUrl]]]];
}
function number(value) {
  if (Number.isNaN(value)) {
    return [];
  }
  return [[value.toString()]];
}
function date(dateString) {
  validateDateString(dateString);
  const dateValue = {
    type: "date",
    start_date: dateString
  };
  return createDateToken(dateValue);
}
function dateTime(date2, timeZone) {
  const dateValue = {
    type: "datetime",
    start_date: formatDate(date2),
    start_time: formatTime(date2)
  };
  if (timeZone) {
    dateValue.time_zone = timeZone;
  }
  return createDateToken(dateValue);
}
function dateRange(startDate, endDate) {
  validateDateString(startDate);
  validateDateString(endDate);
  const dateValue = {
    type: "daterange",
    start_date: startDate,
    end_date: endDate
  };
  return createDateToken(dateValue);
}
function dateTimeRange(startDate, endDate, timeZone) {
  const dateValue = {
    type: "datetimerange",
    start_date: formatDate(startDate),
    start_time: formatTime(startDate),
    end_date: formatDate(endDate),
    end_time: formatTime(endDate)
  };
  if (timeZone) {
    dateValue.time_zone = timeZone;
  }
  return createDateToken(dateValue);
}
function link(displayText, url2) {
  return [[displayText, [["a", url2]]]];
}
function select(value) {
  return [[value]];
}
function multiSelect(...values) {
  if (values.length === 0) {
    return [];
  }
  return [[values.join(",")]];
}
function status(value) {
  return [[value]];
}
function people(...emails) {
  return emails.map((email2) => ({ email: email2 }));
}
function place(value) {
  if (typeof value.lat !== "number" || typeof value.lon !== "number") {
    throw new Error("Place value must have numeric lat and lon coordinates");
  }
  return value;
}
function relation(primaryKey) {
  return { type: "primaryKey", value: primaryKey };
}
function validateDateString(dateString) {
  if (!/^\d{4}-\d{2}-\d{2}$/.test(dateString)) {
    throw new Error(
      `Invalid date format: ${dateString}. Expected YYYY-MM-DD format.`
    );
  }
  const date2 = new Date(dateString);
  if (Number.isNaN(date2.getTime())) {
    throw new Error(`Invalid date: ${dateString}`);
  }
}
function createDateToken(dateValue) {
  return [["\u2023", [["d", dateValue]]]];
}
function formatDate(date2) {
  const year = date2.getFullYear();
  const month = String(date2.getMonth() + 1).padStart(2, "0");
  const day = String(date2.getDate()).padStart(2, "0");
  return `${year}-${month}-${day}`;
}
function formatTime(date2) {
  const hours = String(date2.getHours()).padStart(2, "0");
  const minutes = String(date2.getMinutes()).padStart(2, "0");
  return `${hours}:${minutes}`;
}
function emojiIcon(emoji) {
  return {
    type: "emoji",
    value: emoji
  };
}
function notionIcon(icon, color = "gray") {
  return {
    type: "notion",
    icon,
    color
  };
}
function imageIcon(url2) {
  return {
    type: "image",
    url: url2
  };
}
export {
  checkbox,
  date,
  dateRange,
  dateTime,
  dateTimeRange,
  email,
  emojiIcon,
  file,
  imageIcon,
  link,
  multiSelect,
  notionIcon,
  number,
  people,
  phoneNumber,
  place,
  relation,
  richText,
  select,
  status,
  text,
  title,
  url
};
